openverse_catalog/dags/providers/provider_dag_factory.py

```diff
@@ -276,82 +256,134 @@ def create_provider_api_workflow(
+        dagrun_timeout=conf.dagrun_timeout,
+        schedule_interval=conf.schedule_string,
+        start_date=conf.start_date,
+        catchup=False,
+        doc_md=conf.doc_md,
+        tags=["provider-reingestion"],
+        render_template_as_native_obj=True,
+        user_defined_macros={"date_partition_for_prefix": date_partition_for_prefix},
+    )
+    with dag:
+        # Generate a list of lists of ingestion TaskGroups for each day of reingestion.
+        ingest_operator_list_list, ingestion_metrics = _build_ingest_operator_list_list(
+            reingestion_day_list_list, conf
+        )
+
+        # For each 'level', make a wait task that waits for all of the reingestion taks
+        # at that level to complete.
+        for i in range(len(ingest_operator_list_list) - 1):
+            wait_operator = EmptyOperator(
+                task_id=f"wait_L{i}", trigger_rule=TriggerRule.ALL_DONE
+            )
+
+            # Set the waiter downstream of all ingestion TaskGroups in the ith list.
+            cross_downstream(ingest_operator_list_list[i], [wait_operator])
+
+            # Set the waiter upstream of all ingestion TaskGroups in the i+1th list.
+            # This gates the tasks behind the waiter.
+            wait_operator >> ingest_operator_list_list[i + 1]
+        ingest_operator_list_list[-1]
+
+        # Create a single report_load_completion task, passing in the list of duration
+        # and counts data for each completed task.
+        report_load_completion = create_report_load_completion(
+            conf.dag_id, conf.media_types, ingestion_metrics, conf.dated
+        )
+
+        # report_load_completion is downstream of all the ingestion TaskGroups in the
+        # final list.
+        cross_downstream(ingest_operator_list_list[-1], [report_load_completion])
+
+    return dag
 
-    Optional Arguments:
 
-    start_date:              datetime.datetime giving the
-                             first valid execution_date of the DAG.
-    max_active_tasks:             integer that sets the number of tasks which
-                             can run simultaneously for this DAG. It's
-                             important to keep the rate limits of the
-                             Provider API in mind when setting this
-                             parameter.
-    default_args:            dictionary which is passed to the
-                             airflow.dag.DAG __init__ method and used to
-                             optionally override the DAG_DEFAULT_ARGS.
-    dagrun_timeout:          datetime.timedelta giving the total amount
-                             of time a given dagrun may take.
-    ingestion_task_timeout:  datetime.timedelta giving the amount of
-                             time a call to the `main_function` is
-                             allowed to take.
+def _build_ingest_operator_list_list(
+    reingestion_day_list_list: list[list[int]], conf: ProviderReingestionWorkflow
```

Comment:
I am not the biggest fan of `_list_list` 😅 Could we maybe do `partitioned_reingestion_days` or an equivalent? The typing makes it clear it should be a list of lists. Same for other `_list_list` occurrences below.

Reply:
Totally -- I had the hardest time coming up with a decent name, but I like `partitioned_reingestion_days`! Thanks 💯